lib/cobra_commander/calculated_component_tree.rb

```diff
@@ -0,0 +1,146 @@
+# frozen_string_literal: true
+
+require "cobra_commander/component_tree"
+
+module CobraCommander
+  # Represents a dependency tree in a given context
+  class CalculatedComponentTree < ComponentTree
+    attr_reader :name, :path
+
+    def initialize(name, path, ancestry = Set.new)
+      super(name, path)
+      @ancestry = ancestry
+      @ruby = Ruby.new(path)
+      @js = Js.new(path)
+      @type = type_of_component
+    end
+
+    def dependencies
+      @deps ||= begin
+        deps = @ruby.dependencies + @js.dependencies
+        deps.sort_by { |dep| dep[:name] }
+            .map(&method(:dep_representation))
+      end
+    end
+
+  private
+
+    def type_of_component
+      return "Ruby & JS" if @ruby.gem? && @js.node?
+      return "Ruby" if @ruby.gem?
+      return "JS" if @js.node?
+    end
+
+    def dep_representation(dep)
+      full_path = File.expand_path(File.join(path, dep[:path]))
+      ancestry = @ancestry + [{ name: @name, path: path, type: @type }]
+      CalculatedComponentTree.new(dep[:name], full_path, ancestry)
+    end
+
+    # Calculates ruby dependencies
+    class Ruby
+      def initialize(root_path)
+        @root_path = root_path
+      end
+
+      def dependencies
+        @deps ||= begin
+          return [] unless gem?
+          gems = bundler_definition.dependencies.select { |dep| path?(dep.source) }
+          format(gems)
+        end
+      end
+
+      def path?(source)
+        return if source.nil?
+        source_has_path = source.respond_to?(:path?) ? source.path? : source.is_a_path?
+        source_has_path && source.path.to_s != "."
+      end
+
+      def format(deps)
+        deps.map do |dep|
+          path = File.join(dep.source.path, dep.name)
+          { name: dep.name, path: path }
+        end
+      end
+
+      def gem?
+        @gem ||= File.exist?(gemfile_path)
+      end
+
+      def bundler_definition
+        ::Bundler::Definition.build(gemfile_path, gemfile_lock_path, nil)
+      end
+
+      def gemfile_path
+        File.join(@root_path, "Gemfile")
+      end
+
+      def gemfile_lock_path
+        File.join(@root_path, "Gemfile.lock")
+      end
+    end
+
+    # Calculates js dependencies
+    class Js
```

Comment:
should we consider changing the names of these internal classes? I think they are too generic. Something like `JsDepencenciesCalculator`

Reply:
Maybe in a separate refactoring. I don't like to include changes that are unnecessary to this implementation.